packages/common/src/hooks/useHarmonyField.ts

```diff
@@ -0,0 +1,262 @@
+ * on the event target in onChange and onBlur, instead using the passed in
+ * field name path. This makes usage with ReactNative more straightforward.
+ *
+ * @param propsOrFieldName the field's name, or the full props for the hook.
+ * @example
+ * const MyComponent = () => {
+ *    const [emailField, { error }] = useHarmonyField({ name: 'email', label: 'Email' })
+ *    return <TextInput {...field} error={!!error} helperText={error} />
+ * }
+ */
+export const useHarmonyField = <Value = any>(
+  propsOrFieldName: string | UseHarmonyFieldProps<Value>
+): [
+  FormikFieldInputProps<Value>,
+  FieldMetaProps<Value>,
+  FieldHelperProps<Value>
+] => {
+  const props =
+    typeof propsOrFieldName === 'string'
+      ? { name: propsOrFieldName }
+      : propsOrFieldName
+  const {
+    name,
+    type = 'text',
+    debouncedIdleMs = 0,
+    transformValueOnChange,
+    transformValueOnBlur,
+    showErrorBeforeSubmit = false,
+    validateFieldOnChange
+  } = props
+  const [field, meta, helpers] = useFormikField(props)
+  const { touched, error } = meta
+  const { setValue, setTouched } = helpers
+  const {
+    validateField,
+    validateOnBlur,
+    validateOnChange: validateFormOnChange,
+    submitCount
+  } = useFormikContext()
+
+  // Tracks whether the user is actively editing the field
+  const [isChanging, setIsChanging] = useState(false)
+
+  // Combine the form validateOnChange prop with this field's prop
+  const validateOnChange =
+    validateFieldOnChange === undefined
+      ? validateFormOnChange
+      : validateFieldOnChange
+
+  // Only show an error state if:
+  // - There's an error in validation, AND
+  // - The field has been touched (focused at least once), AND
+  // - The field can error before submission or the form has been submitted at least once, AND
+  // - The field is not currently being changed
+  const hasError = Boolean(
+    error &&
+      touched &&
+      (showErrorBeforeSubmit || submitCount > 0) &&
+      !isChanging
+  )
+
+  /**
+   * Sets the field to touched, validates it, and then sets isChanging to false.
+   * Essentially mimicking onBlur but ensuring validation finishes before
+   * allowing errors to show.
+   */
+  const validateAndSetIdle = useCallback(() => {
+    // We'll validate ourselves - pass false for validation
+    setTouched(true, false)
+    // Cast back to promise (the types are wrong)
+    const maybePromise = validateField(name) as Promise<any> | void
+    // Wait for validation before confirming we're not editing
+    if (maybePromise) {
+      maybePromise.then(() => {
+        setIsChanging(false)
+      })
+    } else {
+      setIsChanging(false)
+    }
+  }, [name, validateField, setIsChanging, setTouched])
+
+  // Debounced validateAndSetIdle for when the user stops typing for a bit
+  const debouncedValidateAndSetIdle = useDebouncedCallback(
+    validateAndSetIdle,
+    [validateAndSetIdle],
+    debouncedIdleMs
+  )
+
+  const onChangeText = useCallback(
```

Comment:
Feels a little strange to have onChangeText defined generally for all field types?

Edit: I see now that we do the check later on :)